myDna pull request

skorochkin opened the pull request.

skorochkin committed:
myDna project added

skorochkin committed:
added initWithCapacity and percentage calculation

skorochkin committed:
additional code comments

iskachkov commented on myDna/myDna/Cell.m:
```diff
@@ -0,0 +1,78 @@
+    NSLog(@"%@", DNA);
+}
+
+-(NSMutableArray *) getDNA
```
Почему бы не использовать property?

skorochkin replied:
Окей, пожалуй переключусь на использование проперти (хотя после лекций/документации в инете - не видел особого смысла использовать - геттер я сделал свой, а памятью в любом случае управляет автоматически что в случае проперти что в случае c alloc/init)
В любом случае, спасибо за замечание :)

iskachkov commented on myDna/myDna/Cell.m:
```diff
@@ -0,0 +1,78 @@
+    
+    for (NSInteger i=0; i<100; ++i) {
+        // check if our array element equals otherDNA element
+        if ([DNA objectAtIndex:i] == [otherDNA objectAtIndex:i]) {
```
Здесь ошибка. Надо считать количество неравных объектов.

skorochkin replied:
Спасибо, исправил в следующем коммите :)

iskachkov commented on myDna/myDna/main.m:
```diff
@@ -0,0 +1,63 @@
+    NSMutableArray *uniqNumbers = [[NSMutableArray alloc] init];
+    
+    // Calculate items count according to percent
+    while ([uniqNumbers count] < (int)round([DNA count] * percent / 100)) {
```
Мне кажется такая конструкция неоптимальной. Могу ошибаться, но, кажется, что каждый цикл будет перелопачиваться весь массив для подсчета количества элементов в нем.

skorochkin replied:
В этой ситуации while будет отрабатывать до тех пор, пока количеством элементов в нем не будет равно проценту от  всех элементов DNA, т.е. через (int)round([DNA count] \* percent / 100) итераций.

iskachkov commented on myDna/myDna/main.m:
```diff
@@ -0,0 +1,63 @@
+            [uniqNumbers addObject: [NSNumber numberWithInteger:r]];
+            
+            // Randomize choosen DNA element
+            [DNA setObject:[self getRandom] atIndexedSubscript:r];
```
Возможна ситуация, когда элемент на самом деле не заменится. Например, C на такой же элемент C.

skorochkin replied:
Добавил обработчик, спасибо!

iskachkov commented on myDna/myDna/main.m:
```diff
@@ -0,0 +1,63 @@
+        
+        // Declare new Cell class instance
+        Cell *myCell1 = [[Cell alloc] initWithCapacity:243];
+        Cell *myCell2 = [[Cell alloc] initWithCapacity:243];
```
По условию задачи емкость ДНК должна быть 100.

skorochkin replied:
Оу, спасибо, забыл переключить на обычный init после тестирования. Уже исправлено.

skorochkin committed:
FINAL

- switched to property usage
- hammingDistance fix
- additional checks in mutation method
-

skorochkin committed:
FINAL

- small fix
